### ISSUE TITLE
ShadowRoot.styleSheets should return an empty StyleSheetList if the shadow root is not connected

### ISSUE THREAD

**Issue by hayatoito** [WICG/webcomponents#526]
It looks that the current spec has an unintentional bug in `ShadowRoot.styleSheets`. To return a meaningful StyleSheetList, UA has to parse (the contents of) style elements even if the shadow root is not in a shadow-including document.

That is inconsistent with the current Web, I think. Processing a style element should be delayed until the style element is connected to a document.

The spec should clarify that it is okay to return an empty StyleSheetList if the shadow root is not in a shadow-including document.

**Comment by hayatoito** [comment]
I will file an issue to HTML Standard too.

**Comment by rniwa** [comment]
This was intentional per discussion on https://github.com/w3c/webcomponents/issues/56.

**Comment by rniwa** [comment]
@annevk @travisleithead

**Comment by hayatoito** [comment]
Ops. I did not realize that we had a discussion. It looks I had actually closed the issue without thinking crefully. :(
Also I found this: https://github.com/whatwg/html/issues/1194

It looks this is an intentional change.

However, I think we should revert this https://github.com/whatwg/html/pull/1199

The reason is :
I am afraid that most UAs can not return `StyleSheetList` for disconnected shadow trees because each item of the list should be `CSSStyleRule`, which has a property `rules`
which should be `CSSRuleList`. We can not return CSSRuleList because we have not parsed the contents of stylesheets yet. That's an important optimization. At least, Blink is not constructing `CSSRuleList` for style elements in disconnected trees.

I am not sure how other UAs are doing. But it would make sense to skip this process, as an optimization, I think.

If my understanding is wrong, please correct me.

**Comment by rniwa** [comment]
The behavior here is that stylesheets would continue to not load while it's detached from a tree.  And load  it once it's inserted into either a shadow tree or a document tree.  This is consistent with how a view-less document work as depicted in https://github.com/w3c/webcomponents/issues/56#issuecomment-194421993.

I would be **strongly** opposed to changing the behavior here.  Please initiate a call for consensus if you'd like to change the behavior.

**Comment by hayatoito** [comment]
I am confused... It sounds that I and you have the same opinion.

> The behavior here is that stylesheets would continue to not load while it's detached from a tree. And load it once it's inserted into either a shadow tree or a document tree. This is quite consistent with how a view-less document work as depicted in #56 (comment).

Yeah, that is exactly what I am proposing here.

**Comment by hayatoito** [comment]
Oh. I understand.
- Your opinion is: The style should be loaded when it is inserted into a shadow tree, even if the shadow tree is not connected to a document.

Could you tell us what's the benefit of this behavior?
- My opinion is: The style should be loaded if and only if it is connected to a document.

**Comment by rniwa** [comment]
Right, this is because a window-less document would still load stylesheets. Since `ShadowRoot` and `Document` both have `styleSheets`, it's inconsistent for `styleSheets` on `ShadowRoot` to behave differently based on whether it's connected to a document or not.  That's just some arbitrary restriction that merits nothing.

**Comment by hayatoito** [comment]
Thanks. Now I think I understand your opinion exactly.
I, however, still think my opinion is more engine friendly. How are WebKit doing here? Do you have any plan?

Note that I am not saying that window-less document should not load style sheets. It's in a shadow-including _document_, I think. So it should load stylesheets. I think we are in the same boat regarding this,

**Comment by rniwa** [comment]
I don't understand what "engine friendly" even means. Given browser engines have the capability to load stylesheets per document, I see no reason they can't do the same per shadow root basis.  As for WebKit's behavior, we don't currently support `styleSheets` on `ShadowRoot`.

**Comment by rniwa** [comment]
I'm pretty sure we resolved this by saying `styleSheets` is empty until we connect it to a document.

This needs to be spec'ed in https://drafts.csswg.org/cssom/#extensions-to-the-document-or-shadow-root-interface

**Comment by rniwa** [comment]
This is basically https://github.com/whatwg/html/issues/4547

**Comment by emilio** [comment]
And https://github.com/w3c/csswg-drafts/issues/3096
